src/dcd/server/autocomplete/ufcs.d

```diff
@@ -0,0 +1,241 @@
+module dcd.server.autocomplete.ufcs;
+
+import dcd.server.autocomplete.util;
+import dsymbol.symbol;
+import dsymbol.scope_;
+import dcd.common.messages;
+import std.functional : unaryFun;
+import std.algorithm;
+import std.array;
+import std.range;
+import dsymbol.builtin.names;
+import std.string;
+import dparse.lexer : tok;
+import std.regex;
+import dcd.server.autocomplete.calltip_utils;
+
+import std.experimental.logger;
+
+void lookupUFCS(Scope* completionScope, DSymbol* beforeDotSymbol, size_t cursorPosition, ref AutocompleteResponse response)
+{
+    // UFCS completion
+    DSymbol*[] ufcsSymbols = getSymbolsForUFCS(completionScope, beforeDotSymbol, cursorPosition);
+
+    foreach (const symbol; ufcsSymbols)
+    {
+        // Filtering only those that match with type of the beforeDotSymbol
+        // We use the calltip since we need more data from dsymbol
+        // hopefully this is solved in the future
+        if (getFirstArgumentOfFunction(symbol.callTip) == beforeDotSymbol.name)
+        {
+            response.completions ~= AutocompleteResponse.Completion(symbol.name, symbol.kind, removeFirstArgumentOfFunction(
+                    symbol.callTip), symbol
+                    .symbolFile, symbol
+                    .location, symbol
+                    .doc);
+        }
+    }
+}
+
+/**
+ * Get symbols suitable for UFCS.
+ *
+ * a symbol is suitable for UFCS if it satisfies the following:
+ * $(UL
+ *  $(LI is global or imported)
+ *  $(LI is callable with $(D implicitArg) as it's first argument)
+ * )
+ *
+ * Params:
+ *     completionScope = current scope
+ *     beforeDotSymbol = the symbol before the dot (implicit first argument to UFCS function)
+ *     cursorPosition = current position
+ * Returns:
+ *     callable an array of symbols suitable for UFCS at $(D cursorPosition)
+ */
+DSymbol*[] getSymbolsForUFCS(Scope* completionScope, const(DSymbol)* beforeDotSymbol, size_t cursorPosition)
+{
+    import containers.hashset : HashSet;
+
+    assert(beforeDotSymbol);
+
+    if (beforeDotSymbol.name is getBuiltinTypeName(tok!"void")
+        || (beforeDotSymbol.type !is null
+            && beforeDotSymbol.type.name is getBuiltinTypeName(tok!"void")))
+    {
+
+        return null; // no UFCS for void
+    }
+
+    Scope* scop = completionScope.getScopeByCursor(cursorPosition);
+    assert(scop);
+    HashSet!size_t visited;
+    // local imports only
+    FilteredAppender!(a => isCallableWithArg(a, beforeDotSymbol), DSymbol*[]) app;
+    while (scop !is null && scop.parent !is null)
+    {
+        auto localImports = scop.symbols.filter!(a => a.kind == CompletionKind.importSymbol);
+        foreach (sym; localImports)
+        {
+            if (sym.type is null)
```

Comment:
Maybe remove braces from if and if-else.

Reply:
alright